### PR TITLE
[Model Averaging] Periodic model averager

### PR DESCRIPTION
**Pull request by wayi1**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#60320 [Model Averaging] Periodic model averager**
* #60303 [Model Averaging] Provide a util function for model averaging

This averager can be used for post-local SGD.

Proposal: https://github.com/pytorch/pytorch/issues/59699
Differential Revision: [D29249850](https://our.internmc.facebook.com/intern/diff/D29249850/)

### CODE REVIEW

File: torch/testing/_internal/distributed/distributed_test.py [hunk 1]
```diff
@@ -976,6 +978,33 @@ def test_average_parameters(self):
...
+            world_size = dist.get_world_size()
+
+            model = nn.Linear(1, 5, bias=False).cuda(device_id)
+            param = next(model.parameters())
+            tensor = torch.ones_like(param.data) * ((rank + 1) // 2)
+            averager = averagers.PeriodicModelAverager(model, warmup_steps=10, period=4)
+            for step in range(0, 20):
+                # Reset the parameters at every step.
+                param.data = copy.deepcopy(tensor)
+                averager.average_parameters(step)
```
Comment (rohan-varma) [inline torch/testing/_internal/distributed/distributed_test.py:1000]: I wonder if we can later enhance the API so that the user doesn't have to call `averager.average_parameters` every iteration. For example, can we register a callback on the DDP module that fires every time the parameters need to be averaged?

File: torch/distributed/algorithms/model_averaging/averagers.py [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+                "of all-reducing parameters will be no less than the cost of all-reducing gradients "
+                "by DistributedDataParall in the backward pass. Therefore, only "
+                "DistributedDataParallel should be used for this case."
+            )
+        self.period = period
+        self.process_group = (
+            process_group if process_group is not None else dist.group.WORLD
+        )
+
+    def average_parameters(self, step: int):
```
Comment (rohan-varma) [inline torch/distributed/algorithms/model_averaging/averagers.py:82]: Can we take `step` out of the API and handle accounting for this ourselves? Longer term maybe eliminate this completely and use a hook as I mentioned in another comment

File: torch/testing/_internal/distributed/distributed_test.py [hunk 3]
```diff
@@ -976,6 +978,33 @@ def test_average_parameters(self):
                 for p in model.parameters():
                     self.assertEqual(p.data, torch.ones_like(p.data) * 0.5)
 
+        @unittest.skipIf(
+            BACKEND != "nccl" and BACKEND != "gloo",
+            "MPI backend does not support creating subgroups on CUDA devices",
+        )
+        @skip_if_lt_x_gpu(2)
+        @skip_if_odd_num_of_gpus()
```
Comment (rohan-varma) [inline torch/testing/_internal/distributed/distributed_test.py:986]: Why does this not work if there is a world size of 5 for example? Can we enhance the test to make it work regardless of world size?

File: torch/testing/_internal/distributed/distributed_test.py [hunk 4]
```diff
@@ -976,6 +978,33 @@ def test_average_parameters(self):
...
+        def test_periodic_model_averager(self):
+            rank = dist.get_rank()
+            rank_to_GPU = self._init_multigpu_helper()
+            device_id = rank_to_GPU[rank][0]
+            world_size = dist.get_world_size()
+
+            model = nn.Linear(1, 5, bias=False).cuda(device_id)
+            param = next(model.parameters())
```
Comment (rohan-varma) [inline torch/testing/_internal/distributed/distributed_test.py:994]: any reason we check only one param vs the entire `model.parameters`?

File: torch/testing/_internal/distributed/distributed_test.py [hunk 5]
```diff
@@ -976,6 +978,33 @@ def test_average_parameters(self):
...
+            tensor = torch.ones_like(param.data) * ((rank + 1) // 2)
+            averager = averagers.PeriodicModelAverager(model, warmup_steps=10, period=4)
+            for step in range(0, 20):
+                # Reset the parameters at every step.
+                param.data = copy.deepcopy(tensor)
+                averager.average_parameters(step)
+                if step < 10 or step % 4 == 0:
+                    # After the model averaging is 0.5
+                    self.assertEqual(param.data, torch.ones_like(param.data) * 0.5)
```
Comment (rohan-varma) [inline torch/testing/_internal/distributed/distributed_test.py:1003]: Does it work for larger groups? 

If world_size = 4, then tensor will be 0.5, 1, 1.5, 2 on respective ranks and the average would be 1.25?

File: torch/distributed/algorithms/model_averaging/averagers.py [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+        >>>  import torch.distributed .algorithms.model_averaging.averagers as averagers
+        >>>  import torch.nn as nn
+        >>>
+        >>>  dist.init_process_group("nccl", rank=rank, world_size=16)
+        >>>  torch.cuda.set_device(rank)
+        >>>  module = nn.Linear(1, 1, bias=False).to(rank)
+        >>>  subgroup, subgroups = dist.new_subgroups()
+        >>>  # Gradients are averaged by each intra-node subgroup during the backward pass.
+        >>>  model = nn.parallel.DistributedDataParallel(
+        >>>     module, device_ids=[rank], output_device=rank, process_group=subgroup
```
[the remark this answers is not in the document]
Comment (SciPioneer) [reply torch/distributed/algorithms/model_averaging/averagers.py:38]: Not really, DDP only run allreduce within each node via the subgroup, so there is no global allreduce. In contrast, `averager` runs global allreduce across nodes.

File: torch/testing/_internal/distributed/distributed_test.py (hunk 1 above)
Comment (SciPioneer) [reply torch/testing/_internal/distributed/distributed_test.py:1000]: It's hard, unless we move optimizer inside DDP. This is because model averaging must occur after `optimizer.step()`, which is actually out of DDP scope. It's a hard boundary at this time.

File: torch/distributed/algorithms/model_averaging/averagers.py (hunk 2 above)
Comment (SciPioneer) [reply torch/distributed/algorithms/model_averaging/averagers.py:82]: Are you suggesting add an internal counter which has exactly the same increment as the `step` in the training loop?

I think this is a good suggestion. I have seen some training loops that were implemented across multiple files, so `step` variable may not be very obviously traceable -- to pass `step` to the averager, it may cause some unnecessary user code change.

Let me update the API in a separate PR.

File: torch/testing/_internal/distributed/distributed_test.py (hunk 3 above)
Comment (SciPioneer) [reply torch/testing/_internal/distributed/distributed_test.py:986]: Yeah, actually I forgot to label this PR [WIP]. I plan to remove `skip_if_odd_num_of_gpus` decorator and simplify the test code.

File: torch/testing/_internal/distributed/distributed_test.py (hunk 4 above)
Comment (SciPioneer) [reply torch/testing/_internal/distributed/distributed_test.py:994]: This is because there is only a single linear layer, so there is only a single parameter tensor.

File: torch/testing/_internal/distributed/distributed_test.py (hunk 5 above)
Comment (SciPioneer) [reply torch/testing/_internal/distributed/distributed_test.py:1003]: Previously I set the value as `(rank + 1) // 2`, so the value will only be 0 or 1.

Anyway, I updated the test to make it easier to understand.

File: torch/testing/_internal/distributed/distributed_test.py [hunk 7]
```diff
@@ -981,6 +982,32 @@ def test_average_parameters(self):
...
+        @skip_if_lt_x_gpu(2)
+        def test_periodic_model_averager(self):
+            rank = dist.get_rank()
+            rank_to_GPU = self._init_multigpu_helper()
+            device_id = rank_to_GPU[rank][0]
+            world_size = dist.get_world_size()
+
+            model = nn.Linear(1, 5, bias=False).cuda(device_id)
+            param = next(model.parameters())
+            tensor = torch.ones_like(param.data) * rank
```
Comment (vadimkantorov) [inline torch/testing/_internal/distributed/distributed_test.py:998]: should this be just `torch.ones_like(param)`?

File: torch/testing/_internal/distributed/distributed_test.py [hunk 8]
```diff
@@ -981,6 +982,32 @@ def test_average_parameters(self):
...
+            expected_avg_tensor = torch.ones_like(param.data) * sum(range(world_size)) / world_size
```
Comment (vadimkantorov) [inline torch/testing/_internal/distributed/distributed_test.py:999]: same here: `torch.ones_like(param)`?

File: torch/testing/_internal/distributed/distributed_test.py (hunk 7 above)
Comment (SciPioneer) [reply torch/testing/_internal/distributed/distributed_test.py:998]: `parameter` is not a normal tensor, but has  both `.data` and `.grad` fields.

**Review by rohan-varma — APPROVED** [review]
Awesome!

File: torch/distributed/algorithms/model_averaging/averagers.py (hunk 2 above)
Comment (SciPioneer) [reply torch/distributed/algorithms/model_averaging/averagers.py:84]: Addressed in #60320
